### PR TITLE
Use pipelines for SocketOutput 

### PR DESCRIPTION
**Pull request by davidfowl**
- Changed socket output to be based on pipelines
- Changed connection filter glue to be based on pipelines
- Codegen that used `MemoryPoolIterator` for output now uses `WritableBuffer`
- Made `UvWriteReq` async/await friendly with `LibuvAwaitable<T>`
- Deleted MemoryPool and friends

TODO:
- ~~Cancellation isn't handled, it needs to be.~~ Will do this in another PR once we get https://github.com/dotnet/corefxlab/pull/1332
- ~~Little bit cleanup~~
- ~~Port relevant `SocketOutput` tests (they are commented out now)~~
- Initial performance test

/cc @pakrym

### CODE REVIEW

File: samples/SampleApp/Startup.cs [hunk 1]
```diff
@@ -47,9 +47,9 @@ public static void Main(string[] args)
                     options.Listen(IPAddress.Loopback, 5000, listenOptions =>
                     {
                         // Uncomment the following to enable Nagle's algorithm for this endpoint.
-                        //listenOptions.NoDelay = false;
+                        listenOptions.NoDelay = false;
 
-                        listenOptions.UseConnectionLogging();
+                        // listenOptions.UseConnectionLogging();
```
Comment (davidfowl) [inline samples/SampleApp/Startup.cs]: Will revert

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/ListenerContext.cs [hunk 2]
```diff
@@ -42,14 +42,22 @@ protected UvStreamHandle CreateAcceptSocket()
...
 
+        public PipeOptions LibuvOutputPipeOptions => new PipeOptions
+        {
+            ReaderScheduler = Thread,
+            WriterScheduler = ServiceContext.ThreadPool,
+            MaximumSizeHigh = ServiceContext.ServerOptions.Limits.MaxResponseBufferSize ?? 0,
+            MaximumSizeLow = ServiceContext.ServerOptions.Limits.MaxResponseBufferSize ?? 0
+        };
+
         public PipeOptions AdaptedPipeOptions => new PipeOptions
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/ListenerContext.cs]: This needs some tweaking. We don't want request limits for both pipes in the adapter.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs [hunk 3]
```diff
@@ -86,5 +89,19 @@ public static ArraySegment<byte> GetArray(this Memory<byte> memory)
             }
             return result;
         }
+
+        public static void WriteAscii(this WritableBuffer buffer, string data)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs]: This needs to be optimized /cc @benaadams 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs [hunk 4]
```diff
@@ -86,5 +89,19 @@ public static ArraySegment<byte> GetArray(this Memory<byte> memory)
...
+            buffer.Write(Encoding.ASCII.GetBytes(data));
+        }
+        public static void Write(this WritableBuffer buffer, string data)
+        {
+            buffer.Write(Encoding.UTF8.GetBytes(data));
+        }
+
+        public static void WriteNumeric(this WritableBuffer buffer, ulong number)
+        {
+            buffer.Write(number.ToString());
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs]: Ugh 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Networking/UvWriteReq.cs [hunk 5]
```diff
@@ -83,12 +85,14 @@ public LibuvAwaitable<UvWriteReq> WriteAsync(UvStreamHandle handle, ArraySegment
                 var index = 0;
                 foreach (var memory in buffer)
                 {
-                    var tryGetPointerResult = memory.TryGetPointer(out var pointer);
-                    Debug.Assert(tryGetPointerResult);
+                    // REVIEW: This isn't necessary for our default pool since the memory is 
+                    // already pinned but it also makes tests pass
+                    var memoryHandle = memory.Pin();
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Networking/UvWriteReq.cs]: That test is wrong,  passing `ReadableBuffer.Create` result which is array based and unexpected. We should just change the test.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Networking/UvWriteReq.cs]: It isn't wrong, just unnecessary if we were to use pre pinned buffers. I also think the overhead is negligible

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PipelineExtensions.cs]: wtf?

File: test/shared/TestApplicationErrorLogger.cs [hunk 6]
```diff
@@ -17,7 +16,7 @@ public class TestApplicationErrorLogger : ILogger
 
         public bool ThrowOnCriticalErrors { get; set; } = true;
 
-        public List<LogMessage> Messages { get; } = new List<LogMessage>();
+        public ConcurrentBag<LogMessage> Messages { get; } = new ConcurrentBag<LogMessage>();
```
Comment (benaadams) [inline test/shared/TestApplicationErrorLogger.cs]: Going for eventually consistent model 😆 

ConcurrentQueue?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs [hunk 7]
```diff
@@ -197,19 +198,17 @@ public void Tick(long timestamp)
                 {
                     _filteredStream = adapterContext.ConnectionStream;
                     _adaptedPipeline = new AdaptedPipeline(
-                        ConnectionId,
                         adapterContext.ConnectionStream,
                         Thread.PipelineFactory.Create(ListenerContext.AdaptedPipeOptions),
-                        Thread.Memory,
-                        Log);
+                        Thread.PipelineFactory.Create(ListenerContext.AdaptedPipeOptions));
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: Use named args here.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Networking;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure
+{
+    public class LibuvAwaitable<TRequest> : ICriticalNotifyCompletion where TRequest : UvRequest
+    {
+        private readonly static Action CALLBACK_RAN = () => { };
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs]: WHY CAPS

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs (hunk 7 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: why

File: test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs [hunk 9]
```diff
@@ -1,865 +1,866 @@
...
+//using Microsoft.AspNetCore.Server.Kestrel.Internal;
+//using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+//using Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure;
+//using Microsoft.AspNetCore.Server.KestrelTests.TestHelpers;
+//using Microsoft.AspNetCore.Testing;
+//using Xunit;
+
+//namespace Microsoft.AspNetCore.Server.KestrelTests
+//{
+//    public class SocketOutputTests
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs]: So it works? :trollface:

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/ListenerContext.cs [hunk 10]
```diff
@@ -42,14 +42,22 @@ protected UvStreamHandle CreateAcceptSocket()
...
             MaximumSizeHigh = ServiceContext.ServerOptions.Limits.MaxRequestBufferSize ?? 0,
             MaximumSizeLow = ServiceContext.ServerOptions.Limits.MaxRequestBufferSize ?? 0
         };
 
+        public PipeOptions LibuvOutputPipeOptions => new PipeOptions
+        {
+            ReaderScheduler = Thread,
+            WriterScheduler = ServiceContext.ThreadPool,
+            MaximumSizeHigh = ServiceContext.ServerOptions.Limits.MaxResponseBufferSize ?? 0,
+            MaximumSizeLow = ServiceContext.ServerOptions.Limits.MaxResponseBufferSize ?? 0
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/ListenerContext.cs]: This looks like a new option is needed.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs [hunk 11]
```diff
@@ -546,269 +195,80 @@ Task ISocketOutput.WriteAsync(ArraySegment<byte> buffer, bool chunk, Cancellatio
...
-                        Debug.Assert(Self._head == Self._tail);
-                        Debug.Assert(Self._tail.Start == Self._tail.End);
+                        var writeReq = _writeReqPool.Allocate();
+                        var writeResult = await writeReq.WriteAsync(_socket, buffer);
+                        _writeReqPool.Return(writeReq);
 
-                        Self._head = null;
-                        Self._tail = null;
-                        returnAll = true;
+                        // REVIEW: Locking here
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: ?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Networking;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure
+{
+    public class LibuvAwaitable<TRequest> : ICriticalNotifyCompletion where TRequest : UvRequest
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs]: nit: put `where` clause in it's own line

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+        public void UnsafeOnCompleted(Action continuation)
+        {
+            OnCompleted(continuation);
+        }
+    }
+
+    public struct UvWriteResult
+    {
+        public int Status;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs]: `readonly`

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        private readonly static Action CALLBACK_RAN = () => { };
+
+        private Action _callback;
+
+        private Exception _exception;
+
+        private int _status;
+
+        public static Action<TRequest, int, Exception, object> Callback = (req, status, error, state) =>
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/LibuvAwaitable.cs]: `readonly`

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs (hunk 7 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: Can't tell what is what.

File: src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/AdaptedPipeline.cs [hunk 15]
```diff
@@ -16,30 +14,50 @@ public class AdaptedPipeline : IDisposable
...
-            IPipe pipe,
-            MemoryPool memory,
-            IKestrelTrace logger)
+            IPipe inputPipe,
+            IPipe outputPipe)
         {
-            Input = pipe;
-            Output = new StreamSocketOutput(connectionId, filteredStream, memory, logger);
+            Input = inputPipe;
+            _output = new StreamSocketOutput(filteredStream, outputPipe);
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/AdaptedPipeline.cs]: Merge `StreamSocketOutput` into `AdaptedPipeline `

File: src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/StreamSocketOutput.cs [hunk 16]
```diff
@@ -3,124 +3,130 @@
...
-                         _outputStream.Write(block.Data.Array, block.Data.Offset, block.Data.Count);
+                        if (buffer.IsEmpty && readResult.IsCompleted)
+                        {
+                            break;
+                        }
+
+                        if (buffer.IsEmpty)
+                        {
+                            await _outputStream.FlushAsync();
+                        }
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/StreamSocketOutput.cs]: else?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs [hunk 17]
```diff
@@ -15,285 +15,111 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
...
             CancellationToken cancellationToken,
-            bool chunk = false,
-            bool socketShutdownSend = false,
-            bool socketDisconnect = false,
-            bool isSync = false)
+            bool chunk = false)
         {
-            TaskCompletionSource<object> tcs = null;
-            var scheduleWrite = false;
+            var flushAwaiter = default(WritableBufferAwaitable);
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: WritableBufferAwaitable flushAwaiter;

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs [hunk 18]
```diff
@@ -15,285 +15,111 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
...
+                    // Graceful shutdown
+                    _pipe.Reader.CancelPendingRead();
                     break;
                 case ProduceEndType.SocketDisconnect:
-                    WriteAsync(default(ArraySegment<byte>),
-                        default(CancellationToken),
-                        socketShutdownSend: false,
-                        socketDisconnect: true,
-                        isSync: true);
+                    // Not graceful
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: turn switch to if?

File: src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/AdaptedPipeline.cs (hunk 15 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/AdaptedPipeline.cs]: In the next round

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs (hunk 7 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: We only do that for bools and null args.

**Comment by davidfowl** [comment]
Will make all of the nit feedback in a subsequent commit.

**Comment by davidfowl** [comment]
This is merged

**Comment by cesarbs** [comment]
07cbf7faa95034639ff42df2f2057974dcf8a055

**Pull request closed without merging**